src/Analyser/MutatingScope.php

```diff
@@ -2008,6 +2009,182 @@ private function resolveType(Expr $node): Type
 			if ($this->reflectionProvider->hasConstant($node->name, $this)) {
 				/** @var string $resolvedConstantName */
 				$resolvedConstantName = $this->reflectionProvider->resolveConstantName($node->name, $this);
+				// core, https://www.php.net/manual/en/reserved.constants.php
+				if ($resolvedConstantName === 'PHP_VERSION') {
+					return new IntersectionType([
+						new StringType(),
+						new AccessoryNonEmptyStringType(),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_MAJOR_VERSION') {
+					return IntegerRangeType::fromInterval(5, null);
+				}
+				if ($resolvedConstantName === 'PHP_MINOR_VERSION') {
+					return IntegerRangeType::fromInterval(0, null);
+				}
+				if ($resolvedConstantName === 'PHP_RELEASE_VERSION') {
+					return IntegerRangeType::fromInterval(0, null);
+				}
+				if ($resolvedConstantName === 'PHP_VERSION_ID') {
+					return IntegerRangeType::fromInterval(50207, null);
+				}
+				if ($resolvedConstantName === 'PHP_ZTS') {
+					return new UnionType([
+						new ConstantIntegerType(0),
+						new ConstantIntegerType(1),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_DEBUG') {
+					return new UnionType([
+						new ConstantIntegerType(0),
+						new ConstantIntegerType(1),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_MAXPATHLEN') {
+					return IntegerRangeType::fromInterval(1, null);
+				}
+				if ($resolvedConstantName === 'PHP_OS') {
+					return new IntersectionType([
+						new StringType(),
+						new AccessoryNonEmptyStringType(),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_OS_FAMILY') {
+					return new UnionType([
+						new ConstantStringType('Windows'),
+						new ConstantStringType('BSD'),
+						new ConstantStringType('Darwin'),
+						new ConstantStringType('Solaris'),
+						new ConstantStringType('Linux'),
+						new ConstantStringType('Unknown'),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_SAPI') {
+					return new UnionType([
+						new ConstantStringType('apache'),
+						new ConstantStringType('apache2handler'),
+						new ConstantStringType('cgi'),
+						new ConstantStringType('cli'),
+						new ConstantStringType('cli-server'),
+						new ConstantStringType('embed'),
+						new ConstantStringType('fpm-fcgi'),
+						new ConstantStringType('litespeed'),
+						new ConstantStringType('phpdbg'),
+						new IntersectionType([
+							new StringType(),
+							new AccessoryNonEmptyStringType(),
+						]),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_EOL') {
+					return new UnionType([
+						new ConstantStringType("\n"),
+						new ConstantStringType("\r\n"),
+					]);
+				}
+				if ($resolvedConstantName === 'PHP_INT_MAX') {
+					return PHP_INT_SIZE === 8
+						? new UnionType([new ConstantIntegerType(2147483647), new ConstantIntegerType(9223372036854775807)])
+						: new ConstantIntegerType(2147483647);
```

Comment:
IMO, this should alwas be the union, as we cannot be sure when phpstan is running with 32 bit php, the code under test also will be and vice versa.

```suggestion
					return new UnionType([new ConstantIntegerType(2147483647), new ConstantIntegerType(9223372036854775807)]);
```

Reply:
You're right, but this would then break running phpstan itself on 32bit.
Currently it runs there :/

Reply:
Ahh I see. Lgtm